### PR TITLE
RSC preloading mechanism

### LINKED ISSUE #212: Tracking issue: investigate prefetching from RSC to Client Components

### PR DESCRIPTION
**Pull request by phryneas**
Resolves #212 and also prepares for #248 a bit.

# Previews:

https://apollo-git-32cae2-apollo-client-next-package-integration-tests.vercel.app//rsc/dynamic/PreloadQuery/useSuspenseQuery
https://apollo-git-32cae2-apollo-client-next-package-integration-tests.vercel.app//rsc/dynamic/PreloadQuery/queryRef-useReadQuery
https://apollo-git-32cae2-apollo-client-next-package-integration-tests.vercel.app//rsc/dynamic/PreloadQuery/queryRef-refTest

On each you can call `__APOLLO_CLIENT__[Symbol.for("apollo.suspenseCache")].queryRefs` to investigate the queryRefs (if there are any)

### CODE REVIEW

File: integration-test/package.json
Comment (phryneas) [inline integration-test/package.json:12]: Scripts with a `:` can be called from any child project in a `yarn` monorepo. Big DX improvement for a single char :D

File: packages/client-react-streaming/src/DataTransportAbstraction/WrappedApolloClient.tsx
Comment (phryneas) [inline packages/client-react-streaming/src/DataTransportAbstraction/WrappedApolloClient.tsx:280]: Gotta point out that this part is new.

**Comment by phryneas** [comment]
Quick thought: we need some warning if people add `cookies` or `headers` to `context` here as that could result in secrets to be dumped into the stream on accident.

**Comment by phryneas** [comment]
So here we have some previews:

https://apollo-client-nextjs-experimental-nextjs-app-support-adj73zyc0.vercel.app/rsc/dynamic/PreloadQuery/useSuspenseQuery
https://apollo-client-nextjs-experimental-nextjs-app-support-adj73zyc0.vercel.app/rsc/dynamic/PreloadQuery/queryRef-useReadQuery
https://apollo-client-nextjs-experimental-nextjs-app-support-adj73zyc0.vercel.app/rsc/dynamic/PreloadQuery/queryRef-refTest

On each you can call `__APOLLO_CLIENT__[Symbol.for("apollo.suspenseCache")].queryRefs` to investigate the queryRefs (if there are any)

File: packages/client-react-streaming/src/DataTransportAbstraction/WrappedApolloClient.tsx
Comment (phryneas) [inline packages/client-react-streaming/src/DataTransportAbstraction/WrappedApolloClient.tsx:125]: This had to become a bound method so I could overwrite it in a child and still call `super.onQueryStarted`

**Comment by phryneas** [comment]
Okay, tests in Vercel run fine now, and the error in `latest` is expected, since that AC release is still missing functionality we require.

I'll do another pass on everything here and then we should be ready for review 🎉

File: packages/client-react-streaming/src/registerApolloClient.tsx
Comment (jerelmiller) [inline packages/client-react-streaming/src/registerApolloClient.tsx:41]: [nit] As a small thing, could we be a bit more verbose with the TS name here so its easily recognizable what `AC` is?

How about
```suggestion
  TApolloClient extends Promise<ApolloClient<any>> | ApolloClient<any>,
```
or
```suggestion
  ApolloClientInstance extends Promise<ApolloClient<any>> | ApolloClient<any>,
```

(I personally like the `TApolloClient`, but I'll leave it up to you.

File: packages/client-react-streaming/src/PreloadQuery.tsx
Comment (jerelmiller) [inline packages/client-react-streaming/src/PreloadQuery.tsx:75]: This makes me wonder, should the transported query ref have a `.toPromise` function on it since its part of the public interface on `QueryReference`? We can always stub it and warn when using it, but might be best to at least conform to the public interface. Thoughts?

**Review by jerelmiller — APPROVED** [review]
I had a few questions and a couple minor comments, but I am SO STOKED for this feature. Incredible work 🎉

File: packages/client-react-streaming/src/PreloadQuery.tsx
Comment (phryneas) [reply packages/client-react-streaming/src/PreloadQuery.tsx]: see https://github.com/apollographql/apollo-client-nextjs/compare/ba35077...e64236d

File: integration-test/nextjs/src/app/rsc/dynamic/PreloadQuery/queryRef-refTest/page.tsx
Comment (phryneas) [reply integration-test/nextjs/src/app/rsc/dynamic/PreloadQuery/queryRef-refTest/page.tsx]: [e08acd9](https://github.com/apollographql/apollo-client-nextjs/pull/258/commits/e08acd90d3d0525ce3383bad4e1a7cf265e78a77)

File: integration-test/nextjs/src/shared/delayLink.ts
Comment (phryneas) [reply integration-test/nextjs/src/shared/delayLink.ts:4]: Agreed.

File: packages/client-react-streaming/src/registerApolloClient.tsx
Comment (phryneas) [reply packages/client-react-streaming/src/registerApolloClient.tsx:38]: what about `ApolloClientOrPromise`?

Comment (phryneas) [reply packages/client-react-streaming/src/registerApolloClient.tsx]: [b162c52](https://github.com/apollographql/apollo-client-nextjs/pull/258/commits/b162c52b4eafcede69e7b8388e03c571b1ce8293)

File: packages/client-react-streaming/src/transportedQueryRef.ts
Comment (phryneas) [inline packages/client-react-streaming/src/transportedQueryRef.ts:114]: @jerelmiller should I still do all this?

Comment (jerelmiller) [reply packages/client-react-streaming/src/transportedQueryRef.ts:114]: I left it for `useBackgroundQuery` [here](https://github.com/apollographql/apollo-client/pull/11821/files#diff-61cffa353d7574ee991f11babdad1bca78cdac9b4a2ccc4b8d938abf0b28bc2bR247-R256) to avoid issues where the `queryRef` gets disposed by the auto dispose timeout and the hook re-renders (which was a bug before that change in 3.9.10). I did update that check to use a `setTimeout` though since `dispose` is now again wrapped in a `setTimeout`.

I don't remember exactly where you use this hook, but if you're in the same situation as what I described above, then keeping this makes sense.

File: yarn.lock
Comment (jerelmiller) [inline yarn.lock:52]: We might want to get https://github.com/apollographql/apollo-client/pull/11821 out before we go live with the change in `useTransportedQueryRef` just to make sure its still compatible with the change in that PR.

**Review by jerelmiller — APPROVED** [review]
Love it! Let's test this with https://github.com/apollographql/apollo-client/pull/11821 just to make sure that doesn't break something here as well, but otherwise this is incredible work!

File: yarn.lock
Comment (phryneas) [reply yarn.lock:52]: This wraps around useReadQuery and useQueryRefHandlers, and it can also create a queryRef by reviving a transportedQueryRef. We can reduce these effects once React deduplicates transported objects, but for now they're fixed.

File: packages/client-react-streaming/src/transportedQueryRef.ts
Comment (phryneas) [reply packages/client-react-streaming/src/transportedQueryRef.ts]: This wraps around useReadQuery and useQueryRefHandlers, and it can also create a queryRef by reviving a transportedQueryRef. We can reduce these effects once React deduplicates transported objects, but for now they're fixed.

File: packages/client-react-streaming/src/transportedQueryRef.ts
Comment (phryneas) [inline packages/client-react-streaming/src/transportedQueryRef.ts:124]: @jerelmiller could you please take a final look at this bit before I merge this in?

**Review by jerelmiller — APPROVED** [review]
💯

**Pull request merged**
